asserts/snapasserts/snapasserts.go

```diff
@@ -365,6 +387,49 @@ func SideInfoFromSnapAssertions(snapDecl *asserts.SnapDeclaration, snapRev *asse
 	}
 }
 
+// DeriveComponentSideInfoFromDigestAndSize tries to construct a
+// ComponentSideInfo using digest and size for a component and ID/name for the
+// snap to find the relevant assertions with the information in the given
+// database. It will fail with an asserts.NotFoundError if it cannot find them.
+func DeriveComponentSideInfoFromDigestAndSize(resName, snapName, snapID string, compPath, snapSHA3_384 string, resSize uint64, db Finder) (*snap.ComponentSideInfo, error) {
+	// get relevant assertions and reconstruct metadata
+	headers := map[string]string{
+		"snap-id":           snapID,
+		"resource-name":     resName,
+		"resource-sha3-384": snapSHA3_384,
+	}
+	a, err := db.Find(asserts.SnapResourceRevisionType, headers)
+	if err != nil && !errors.Is(err, &asserts.NotFoundError{}) {
+		return nil, err
+	}
+	if a == nil {
+		// non-default provenance?
+		cands, err := db.FindMany(asserts.SnapResourceRevisionType, headers)
+		if err != nil {
+			return nil, err
+		}
+		if len(cands) != 1 {
+			return nil, fmt.Errorf("safely handling resources with different provenance but same hash not yet supported")
+		}
+		a = cands[0]
+	}
+
+	resRev := a.(*asserts.SnapResourceRevision)
+
+	if resRev.ResourceSize() != resSize {
+		return nil, fmt.Errorf("resource %q does not have the expected size according to signatures (broken or tampered): %d != %d", resName, resSize, resRev.ResourceSize())
+	}
+
```

Comment:
if you compare with the same function for snaps, we need to run crossCheckResourceProvenance  here as well at this point